// @tiptap/extension-collaboration-caret/dist/index.js
// src/collaboration-caret.ts
import { Extension } from "@tiptap/core";
import { Plugin, PluginKey } from "@tiptap/pm/state";
import { defaultSelectionBuilder, yCursorPlugin } from "@tiptap/y-tiptap";
var awarenessStatesToArray = (states) => {
  return Array.from(states.entries()).map(([key, value]) => {
    if (value && value.user) {
      return {
        clientId: key,
        ...value.user
      };
    }
    return {
      clientId: key
    };
  });
};
var defaultOnUpdate = () => null;
var CollaborationCaret = Extension.create({
  name: "collaborationCaret",
  priority: 999,
  addOptions() {
    return {
      provider: null,
      user: {
        name: null,
        color: null
      },
      render: (user) => {
        const cursor = document.createElement("span");
        cursor.classList.add("collaboration-carets__caret");
        cursor.setAttribute("style", `border-color: ${user.color}`);
        const label = document.createElement("div");
        label.classList.add("collaboration-carets__label");
        label.setAttribute("style", `background-color: ${user.color}`);
        label.insertBefore(document.createTextNode(user.name), null);
        cursor.insertBefore(label, null);
        return cursor;
      },
      selectionRender: defaultSelectionBuilder,
      onUpdate: defaultOnUpdate
    };
  },
  onCreate() {
    if (this.options.onUpdate !== defaultOnUpdate) {
      console.warn(
        '[tiptap warn]: DEPRECATED: The "onUpdate" option is deprecated. Please use `editor.storage.collaborationCaret.users` instead. Read more: https://tiptap.dev/api/extensions/collaboration-caret'
      );
    }
    if (!this.options.provider) {
      throw new Error('The "provider" option is required for the CollaborationCaret extension');
    }
  },
  addStorage() {
    return {
      users: []
    };
  },
  addCommands() {
    return {
      updateUser: (attributes) => () => {
        this.options.provider.awareness.setLocalStateField("user", attributes);
        return true;
      },
      user: (attributes) => ({ editor }) => {
        console.warn(
          '[tiptap warn]: DEPRECATED: The "user" command is deprecated. Please use "updateUser" instead. Read more: https://tiptap.dev/api/extensions/collaboration-caret'
        );
        return editor.commands.updateUser(attributes);
      }
    };
  },
  addProseMirrorPlugins() {
    const { provider } = this.options;
    const storage = this.storage;
    const user = this.options.user;
    const awarenessListenerPlugin = new Plugin({
      key: new PluginKey("collaborationCaretAwarenessListener"),
      view: () => {
        const onAwarenessUpdate = () => {
          storage.users = awarenessStatesToArray(provider.awareness.states);
        };
        provider.awareness.setLocalStateField("user", user);
        storage.users = awarenessStatesToArray(provider.awareness.states);
        provider.awareness.on("update", onAwarenessUpdate);
        return {
          destroy: () => {
            provider.awareness.off("update", onAwarenessUpdate);
            storage.users = [];
          }
        };
      }
    });
    return [
      awarenessListenerPlugin,
      yCursorPlugin(provider.awareness, {
        cursorBuilder: this.options.render,
        selectionBuilder: this.options.selectionRender
      })
    ];
  }
});

// src/index.ts
var index_default = CollaborationCaret;
export {
  CollaborationCaret,
  index_default as default
};
